### ISSUE TITLE
Still seeing "error adding host side routes for interface: xxx, error: failed to add route file exists" with calico cni 1.11.1

### ISSUE THREAD

**Issue by msavlani** [projectcalico/calico#1406]
Hi Team,

I am still seeing this error with calico cni 1.11.1
As per #1253 it was supposed to be fixed by calico cni 1.11.1.

Error -
Events:
  FirstSeen     LastSeen        Count   From                    SubObjectPath   Type            Reason          Message
  ---------     --------        -----   ----                    -------------   --------        ------          -------
  11m           11m             1       default-scheduler                       Normal          Scheduled       Successfully assigned test-deploy-1176720686-jzd5n to node1
  11m           11m             1       kubelet, node1                    Warning         FailedSync      Error syncing pod, skipping: failed to "SetupNetwork" for "test-deploy-1176720686-jzd5n_test_ns" with SetupNetworkError: "NetworkPlugin cni failed to set up pod \"test-deploy-1176720686-jzd5n_test_ns\" network: error adding host side routes for interface: cali68c508fe1a5, error: route (Ifindex: 32723, Dst: 172.40.205.147/32, Scope: 253) already exists for an interface other than 'cali68c508fe1a5'"

  11m   11m     1       kubelet, node1            Warning FailedSync      Error syncing pod, skipping: failed to "SetupNetwork" for "test-deploy-1176720686-jzd5n_test_ns" with SetupNetworkError: "NetworkPlugin cni failed to set up pod \"test-deploy-1176720686-jzd5n_test_ns\" network: error adding host side routes for interface: cali68c508fe1a5, error: route (Ifindex: 32726, Dst: 172.40.205.147/32, Scope: 253) already exists for an interface other than 'cali68c508fe1a5'"

As suggested in solution under #1253 , i did rolling restart of all pods running on affected node.
Also please note, i see this error only after restart of statefulset.
That is IP which is being complained above "172.40.205.147" is being used by one of the statefulset pod.
I see this behavior is consistent. i.e. IP under question is always one which is used by STATEFULSET

Environment -

Calico Node : v2.4.1
Calico CNI : v1.11.1
K8s : 1.6.7

**Comment by msavlani** [comment]
@fasaxc @gunjan5

**Comment by fasaxc** [comment]
@msavlani can you delete the stateful set to free the IP and then recreate it?  Once you're in this state, you'll have an IP that IPAM considers to be free but that is in-use by a Pod; in this case a statefulset Pod, which presumably keeps its IP over restart.

**Comment by fasaxc** [comment]
@msavlani How is your stateful set configured, how does it get its IP addresses?

@caseydavenport Do you know how stateful sets interact with CNI/IPAM, is there any special plumbing?

**Comment by msavlani** [comment]
@fasaxc i didnt understand that completely . You mean i need to delete statefulset as whole and not just pods which are part of statefulset ? 
How does that make difference ?
 I have 4 nodes cluster and when i  delete pods from statefulset, k8s bring them up again not necessarily on same host (and thus they get different IP's as per podCIDR for that node). This means i dont get static IP's for statefulset.

**Comment by fasaxc** [comment]
@msavlani Hmm, that should be OK then.  Note: you may have the problem on more than one node with no symptoms. until a new pod is started on the node that has previously hit the problem.  You may need to do a rolling restart on other hosts too (or all your hosts).

Afraid that's the easiest way to clear the bad data, it's pretty tricky to do anything cleaner.

**Comment by Random-Liu** [comment]
@fasaxc How and when do we update calico version in Kubernetes upstream? https://github.com/kubernetes/kubernetes

**Comment by fasaxc** [comment]
@Random-Liu We'll put up a PR to rev calico as soon as we ship 2.6.3, which is in QA at the moment.

**Comment by msavlani** [comment]
@fasaxc I already did rolling restart of all the pods on all the nodes with statefulset being first and then other deployments.
And i still encountered same route exists error when deployment pods were coming up.

**Comment by fasaxc** [comment]
@msavlani Are you sure the new CNI plugin is active?  What steps did you take to upgrade it?

**Comment by caseydavenport** [comment]
> Do you know how stateful sets interact with CNI/IPAM, is there any special plumbing?

@fasaxc There's no special code for stateful sets. The main thing that could be different is that they get a consistent name even when the pod is re-created, but we don't have any special code paths for that.

**Comment by fasaxc** [comment]
@msavlani I'd really like to get to the bottom of this issue.  If you reach out to me on the calicousers slack (you can join here http://slack.projectcalico.org/) then we could do a live debug session to get to the bottom of it.

**Comment by r0bj** [comment]
I'm also still seeing it for statefulsets for calico v2.6.3.

kubelet logs:
https://gist.github.com/r0bj/f5c89fde8e96f0649951acb3289d1ef2
calico-node logs:
https://gist.github.com/r0bj/8a4b282dcdb1d3ab23ceae60abe4b1c3

**Comment by fasaxc** [comment]
@r0bj Thanks for the logs, looks like you're still seeing the IPAM-related issue.  As I mentioned above, the actual bug is that we mark an IP as free when it's still in use so the error happens a long time after the bug (potentially).

Do you have kubelet logs going back any further; it'd be handy to find any mentions of the conflicting IP address being assigned/freed: `10.203.131.144` from before the log where you see the error.  

Another angle we could look at is what workload endpoints you have on that host:
```
calicoctl get workloadendpoint --node=<node> -o yaml
```
should list them and then you can look for any that are sharing IP addresses, in particular the clashing one: `10.203.131.144`

**Comment by r0bj** [comment]
@fasaxc thanks for clarification. Next time I will try to collect required data.

**Comment by fasaxc** [comment]
@r0bj Another option is to drain a whole node of pods so that you get to 0 pods on the node and then reintroduce them.  I'm wondering if you're seeing a problem where one pod is being created while another (with a missing IPAM reservation) is being deleted.  Completely draining a node should remove all the IPAM allocations and allow the new CNI plugin to do the allocations correctly.

**Comment by fasaxc** [comment]
@r0bj @msavlani Also worth looking for these new logs that were added in the new CNI plugin:
```
"Not cleaning up IPAM allocation; this was a pre-existing endpoint"
"Releasing IPAM allocation after failure"
```

**Comment by r0bj** [comment]
@fasaxc This time all pod were recreated. Pod `celery-vcl-tester-7c976cb4df-q5zwk` fail to create.
IP address `10.203.97.156` is already assign to pod `monitoring.alertmanager-main-2` (StatefulSets) so it cannot be used.

Logs are from the beginning after node was restarted.

kubelet logs:
https://gist.github.com/r0bj/e1ffdeb858c1259d1438ea88a2e58f70

calico-node logs:
https://gist.github.com/r0bj/f5a9ca11cae4f8923a2f6d719b3fdae2

output of calicoctl get workloadendpoint --node=dev-k8s-worker-p1 -o yaml:
https://gist.github.com/r0bj/9f1ad752fb2184da3adbc4499701701f

ip route on the node:
https://gist.github.com/r0bj/88b636f7124005efff41b3371efc0a2b

```
2017-12-01 14:17:16.775 [INFO][18519] calico-ipam.go 177: Auto assigning IP Workload="ops.celery-vcl-tester-7c976cb4df-q5zwk" assignArgs=client.AutoAssignArgs{Num4:1, Num6:0, HandleID:(*string)(0xc4204ce940), Attrs:map[string]string(nil), Hostname:"dev-k8s-worker-p1", IPv4Pools:[]net.IPNet{}, IPv6Pools:[]net.IPNet{}}
2017-12-01 14:17:16.775 [INFO][18519] ipam.go 120: Auto-assign 1 ipv4, 0 ipv6 addrs for host 'dev-k8s-worker-p1'
2017-12-01 14:17:16.798 [INFO][18519] ipam.go 484: Block 10.203.100.192/26 is full
2017-12-01 14:17:16.806 [INFO][18519] ipam_block.go 343: New allocation attribute: {AttrPrimary:0xc4204ce940 AttrSecondary:map[]}
2017-12-01 14:17:16.816 [INFO][18519] ipam.go 816: Creating new handle: ops.celery-vcl-tester-7c976cb4df-q5zwk
W1201 14:17:16.824574    1492 helpers.go:847] eviction manager: no observation found for eviction signal allocatableNodeFs.available
2017-12-01 14:17:16.835 [INFO][18519] ipam.go 285: Auto-assigned 1 out of 1 IPv4s: [10.203.97.156]
Calico CNI IPAM assigned addresses IPv4=[10.203.97.156] IPv6=[]
2017-12-01 14:17:16.836 [INFO][18519] calico-ipam.go 205: IPAM Result Workload="ops.celery-vcl-tester-7c976cb4df-q5zwk" result.IPs=[]*current.IPConfig{(*current.IPConfig)(0xc4204aac00)}
2017-12-01 14:17:16.838 [INFO][18488] k8s.go 265: Populated endpoint Workload="ops.celery-vcl-tester-7c976cb4df-q5zwk" endpoint=WorkloadEndpoint(Node=dev-k8s-worker-p1, Orchestrator=k8s, Workload=ops.celery-vcl-tester-7c976cb4df-q5zwk, Name=eth0)
Calico CNI using IPs: [10.203.97.156/32]
I1201 14:17:16.856851    1492 kuberuntime_manager.go:389] No ready sandbox for pod "elastic-exporter-9bcfcdc67-ngxwf_monitoring(540cf934-d6a2-11e7-bc1a-5254009ef6db)" can be found. Need to start a new one
W1201 14:17:16.858556    1492 cni.go:265] CNI failed to retrieve network namespace path: Cannot find network namespace for the terminated container "1e8868423fc0553025fa40b7731c4c5284b52213ea295f098957423863cd7c51"
2017-12-01 14:17:16.890 [ERROR][18488] k8s.go 286: Error setting up networking Workload="ops.celery-vcl-tester-7c976cb4df-q5zwk" error=error adding host side routes for interface: cali2f58c4d4070, error: route (Ifindex: 54, Dst: 10.203.97.156/32, Scope: 253) already exists for an interface other than 'cali2f58c4d4070'
2017-12-01 14:17:16.890 [INFO][18488] k8s.go 278: Releasing IPAM allocation after failure IPs=[]net.IPNet{net.IPNet{IPNet:net.IPNet{IP:net.IP{0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xff, 0xff, 0xa, 0xcb, 0x61, 0x9c}, Mask:net.IPMask{0xff, 0xff, 0xff, 0xff}}}} Workload="ops.celery-vcl-tester-7c976cb4df-q5zwk"
2017-12-01 14:17:16.890 [INFO][18488] utils.go 269: Cleaning up IP allocations for failed ADD IPs=[]net.IPNet{net.IPNet{IPNet:net.IPNet{IP:net.IP{0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xff, 0xff, 0xa, 0xcb, 0x61, 0x9c}, Mask:net.IPMask{0xff, 0xff, 0xff, 0xff}}}} Workload="ops.celery-vcl-tester-7c976cb4df-q5zwk"
2017-12-01 14:17:16.962 [INFO][18541] calico.go 329: Extracted identifiers ContainerID="1e8868423fc0553025fa40b7731c4c5284b52213ea295f098957423863cd7c51" Node="dev-k8s-worker-p1" Orchestrator="k8s" Workload="monitoring.elastic-exporter-9bcfcdc67-ngxwf"
```

```
# /opt/cni/bin/calico -v
v1.11.1
```

**Comment by fasaxc** [comment]
@r0bj I'm seeing some confusing logs, could you send me a dump of your calico etcd data?  This command should do it assuming that etcdctl is on your path and configured to talk to your calico etcd:
```
etcdctl ls --recursive -p /calico | grep -v '/$' | xargs -n1 -I '{}' sh -c "echo; echo {}; etcdctl get {} " | tee etcdctl_get
```
Feel free to DM me on our slack `@shaun.crampton` if you don't want to share the data publicly.

**Comment by r0bj** [comment]
@fasaxc here you have
[etcdctl_get.tar.gz](https://github.com/projectcalico/calico/files/1522149/etcdctl_get.tar.gz)

**Comment by fasaxc** [comment]
Awesome, thanks @r0bj

**Comment by msavlani** [comment]
@fasaxc  did you findout something on this.
I think draining the node will also not solve the problem here because of following.
I see when i start statefulset it gets new IP on node 1.
Now if i start pod say pod1 on node1 , calico-ipam give different IP and everything works fine.
Now if i started new pod say pod2 on node1, calico-ipam someone gives same IP as above statefulset pod and things start failing.

Question#1 is why does calico-ipam gives same IP ?
Question#2 If it gave same IP for some reason and getting errors now, shouldn't we call ipam again to get new IP for affected pod and get the flow moving ?

**Comment by fasaxc** [comment]
@msavlani I haven't figured it out yet, but this is a very high priority bug for us.  As far as I can tell, there's nothing unusual about how a statefulset pod gets networked so I haven't found a good thread to pull on yet.

**Comment by ldnvnbl** [comment]
I had the same problem.  

I find that one ip is used by k8s pod, but calicoctl shows it is not assigned. 


![image](https://user-images.githubusercontent.com/3798479/33645927-3d08ab22-da88-11e7-8698-18a45ff94269.png)
![image](https://user-images.githubusercontent.com/3798479/33645956-6b27ec8e-da88-11e7-99fd-d472dc308e97.png)

**Comment by fasaxc** [comment]
@r0bj We _think_ we've figured out what's going wrong in your case:

- We see IP `10.203.97.156` get assigned to a pod `monitoring.alertmanager-main-2`
- Then, later, it gets assigned to `monitoring.elastic-exporter-9bcfcdc67-ngxwf`
- The etcd dump is missing the IPAM allocation handle for `monitoring.alertmanager-main-2`
- However, the logs show the IPAM handle being created successfully when `monitoring.alertmanager-main-2` is networked.
- Since we know that the handle was created, but the second allocation reuses the IP, and the handle is missing from the etcd dump, it seems likely that the handle was deleted after the first allocation but before the second one.
- IPAM handles are only deleted when a pod with the same name is deleted.
- But,there is no deletion in the log for `monitoring.alertmanager-main-2` between the first allocation and the second
- However, there _are_ lots of deletion _retries_ in the log for `fluentd-node-agent-pr9mf` and other pods.

That leads me to suspect that there is a similar deletion retry happening on another node:

- `monitoring.alertmanager-main-2` is scheduled to node A
- `monitoring.alertmanager-main-2`, or node A is restarted
- node A begins a deletion retry loop for `monitoring.alertmanager-main-2`
- `monitoring.alertmanager-main-2` is rescheduled to node B
- node B successfully creates the IPAM handle for `monitoring.alertmanager-main-2`
- node A retries the deletion (or it may only be a single delayed attempt); sees an IPAM handle called `monitoring.alertmanager-main-2` and cleans it up incorrectly; that IPAM handle belonged to the instance of the pod on node B.

Can you scan your other kubelet logs, if you still have them, for instances of `monitoring.alertmanager-main-2`; would be good to confirm that there's a deletion somewhere that cleaned up the IPAM handle, when it shouldn't.

FWIW, I believe there's a kubelet bug that is making this worse, causing the deletion to be continuously retried.  I believe another user reported that being fixed in 1.8.5 but I haven't been able to track down a Kubernetes issue.

There's a second question, which is why your system is hitting this so reliably.  The issue there appears to be that this particular node has exactly one IP address free in its IPAM block.  That means that it's always trying to re-use the same bad IP for each allocation.  (If the block had 0 IPs left then it would claim a new, fresh block instead, but having exactly one IP left is a corner case.)

Possible workarounds:
- upgrade kubelet, hopefully that will reduce the frequency of this issue (there's still a Calico bug but it'd be much harder to hit if kubelet stops retrying)
- after upgrading kubelet, prevent your stateful set from being scheduled while all the deletions for old instances of the pod are complete, they allow scheduling again

**Comment by r0bj** [comment]
@fasaxc Thanks for explanation.
The only kubelet logs I've found from that time range on different node than `dev-k8s-worker-p1` related to `monitoring.alertmanager-main-2` is this:
https://gist.github.com/r0bj/1f7d765fc939a7a495abeca2127b8ca1

I'm going to test kubelet 1.8.5 when it'll be available (I'm using kubelet 1.8.4 currently).

**Comment by fasaxc** [comment]
@r0bj Great, those logs confirm the hypothesis above:
- Original log: `2017-12-01 14:15:17.111` IP is assigned
- Other node log: `2017-12-01 14:15:58.604 [INFO][19479] ipam.go 738: Releasing all IPs with handle 'monitoring.alertmanager-main-2'`  <- IP gets released incorrectly by the other node
- Original log: `2017-12-01 14:17:15.215` IP gets re-assigned, resulting in failure

**Comment by ldnvnbl** [comment]
@fasaxc How to fix this? 

I have an idea that is to use net container id to create handle that is global unique.

But do this we should think already created pod that created handle use pod name.

Our product k8s cloud use net container id to create handle that is no problem.

**Comment by fasaxc** [comment]
Yes, that's that plan that we've come up with, we're going to switch to use the container ID.

**Comment by fasaxc** [comment]
@heschlie is working on the fix here: https://github.com/projectcalico/cni-plugin/pull/445

**Comment by caseydavenport** [comment]
I think this was fixed in https://github.com/projectcalico/cni-plugin/pull/445

**Comment by fasaxc** [comment]
@r0bj @msavlani Please can you test with this version of CNI plugin: https://github.com/projectcalico/cni-plugin/releases/tag/v1.11.2-rc1

It changes the handle to use the container ID, which should mean that, once any affected pods are restarted, they'll get fresh handles and no-one should remove their IP allocations.

Unfortunately, it doesn't fix if an existing pod has an IP but its IP handle is missing so, if you do see the error again, you need to track down the *old* pod that is still using the IP without a handle and restart it.

**Comment by r0bj** [comment]
@fasaxc I recreated all pods in the cluster (is that enough?) then started test.
Pod `nginx-stateful-test-1` fail to create. Ip address `10.203.100.224/32` is used twice by `ops.nginx-stateful-test-1` and `logstash-k8s-587b7f8d75-mjdcz`.

`logstash-k8s-587b7f8d75-mjdcz` start time:
Start Time:     Fri, 15 Dec 2017 12:38:14 +0000

`nginx-stateful-test-1`start time:
Start Time:     Fri, 15 Dec 2017 12:38:53 +0000

Logs are from the beginning after node was restarted.

kubelet logs:
https://gist.github.com/r0bj/73b543d7a3ee6b36cc369f4c9e20d9cd

calico-node logs:
https://gist.github.com/r0bj/7b35b7fa37d09255d40983923dfeb56a

output of calicoctl get workloadendpoint --node=dev-k8s-worker-p1 -o yaml:
https://gist.github.com/r0bj/7f6d8d1fe937cdecb415a44ed997b130

The same address `10.203.100.224/32` used twice:
```
- apiVersion: v1
  kind: workloadEndpoint
  metadata:
    activeInstanceID: 0381a077d1460d65c0115f35696d2a15b8516dd4830c8c4242bdf14494b3321f
    labels:
      app: logstash-k8s
      calico/k8s_ns: ops
      pod-template-hash: "1436394831"
    name: eth0
    node: dev-k8s-worker-p1
    orchestrator: k8s
    workload: ops.logstash-k8s-587b7f8d75-mjdcz
  spec:
    interfaceName: cali60eade87a56
    ipNetworks:
    - 10.203.100.224/32
    mac: ae:e2:ae:1e:03:96
    profiles:
    - k8s_ns.ops
- apiVersion: v1
  kind: workloadEndpoint
  metadata:
    activeInstanceID: 88c449da8a87661d766a53361a8c82037f1cc296d5d467195210f4e6fc00f54a
    labels:
      app: nginx-stateful-test
      calico/k8s_ns: ops
      controller-revision-hash: nginx-stateful-test-75884d5c6
    name: eth0
    node: dev-k8s-worker-p1
    orchestrator: k8s
    workload: ops.nginx-stateful-test-1
  spec:
    interfaceName: cali0579e1ead8c
    ipNetworks:
    - 10.203.100.224/32
    mac: 5a:b0:a9:e1:de:bc
    profiles:
    - k8s_ns.ops
```

**Comment by fasaxc** [comment]
@r0bj Can you send me an etcd dump?  Let's check the new handle is being used.

**Comment by r0bj** [comment]
@fasaxc This is another case, because I had to fix the previous one.

Failed pod: `nginx-stateful-test-0`
Duplicated IP: 10.203.131.144
Pod with the same IP: `global-discussion-feed-85dbc46d84-zvg4f`

kubelet logs:
https://gist.github.com/r0bj/ede5d0a4c0425c2a8934a01fe1981eff

calico-node logs:
https://gist.github.com/r0bj/73bb26e04fb3253e3efde9dedc59de5a

output of `calicoctl get workloadendpoint --node=dev-k8s-worker-p1 -o yaml`:
https://gist.github.com/r0bj/66062607029434dbcf11f90fca4536d3

etcd dump:
[etcdctl_get.tar.gz](https://github.com/projectcalico/calico/files/1563183/etcdctl_get.tar.gz)

**Comment by fasaxc** [comment]
@r0bj I'm still seeing a lot of old IPAM handles in your data dump, did you restart all your pods on all your nodes?  If not, I recommend doing that.  I think it's possible that some problems may persist if you hit certain timing windows where old pods are being removed and new pods are being added.

When you have two pods with the same IP, I think the safest option is to delete the oldest pod first, so that the pod that has the most recent handle is the last to be deleted.

Note: your cluster also seems to be hitting a kubernets bug where deletions are being retried for pods that are long-dead.  I suggest you upgrade to k8s 1.8.5, which I have heard resolves that issue.  (I'll make the logs a lot easier to follow too, since there won't be hundreds of spurious deletions to filter out.)

**Comment by r0bj** [comment]
@fasaxc All pods were deleted and recreated before I started test.

Age of all pods:
```
# kubectl get pods --all-namespaces --no-headers |awk '{print $6}' | sort -u
1h
4h
5h
```

It is k8s version 1.8.5.

**Comment by fasaxc** [comment]
@r0bj did you update the CNI plugin on all nodes before doing that?

**Comment by r0bj** [comment]
Yes, it's `v1.11.2-rc1` on every node.

**Comment by r0bj** [comment]
I can confirm that behaviour in k8s 1.9.0.

**Comment by fasaxc** [comment]
@r0bj Can you reproduce this issue on a fresh cluster with Calico v2.6.4?  I'm still trying to figure out if there's still an issue or if it's just the problem that was caused by the earlier Calico version still causing problems somehow.

**Comment by caseydavenport** [comment]
What's the latest status on this guy? Have we convinced ourselves this is due to stale data, or do we see this happen on fresh clusters as well?

**Comment by fasaxc** [comment]
@msavlani @r0bj Can you confirm whether you see this on a fresh, clean cluster with the latest Calico?

**Comment by r0bj** [comment]
I still can reproduce it on my existing clusters but unfortunately I didn't have a chance to test it on a fresh cluster.
Is there something to do on my existing clusters (I've recreated all pods) to get objective results?

**Comment by fasaxc** [comment]
@r0bj Can you send me a new etcd dump after recreating all pods?

**Comment by r0bj** [comment]
Sure.
In this case duplicated ip (10.203.131.144) have pods:
`nginx-stateful-test-0`:
```
# kubectl describe pod nginx-stateful-test-0 -n ops
[...]
Start Time:     Thu, 25 Jan 2018 17:13:35 +0000
```

`user-registration-feed-cfbd7bd9b-29hrt`:
```
# kubectl describe pod user-registration-feed-cfbd7bd9b-29hrt -n dev
[...]
Start Time:     Thu, 25 Jan 2018 17:10:43 +0000
```

etcd dump:
[etcdctl_get_20180125.tar.gz](https://github.com/projectcalico/calico/files/1664970/etcdctl_get_20180125.tar.gz)

**Comment by caseydavenport** [comment]
I see a lot of old-style IPAM handles in that output, indicating to me either that some nodes haven't been updated to the latest code or that not all pods have been restarted using the latest code.

I do see _some_ new style handles, which means at least some pods have been started with the new code.

I believe this was the same issue that has been confirmed fixed: https://github.com/projectcalico/calico/issues/1542

**Comment by calder** [comment]
Hitting this with:
* Kubernetes v1.13.3
* Calico v3.1.1

Should we reopen this or should I file a new issue?

**Comment by caseydavenport** [comment]
@calder could you open a new issue? Thanks

**Comment by melwynjensen** [comment]
Is this issue reopened already? If so can you please point the link for the same?

**Comment by fasaxc** [comment]
I think we've fixed various issues that can cause this.  Suggest you upgrade to v3.8 to pick up all the fixes.
